kbncontent.go

```diff
@@ -0,0 +1,314 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// Package kbncontent implements routines for analyzing Kibana content.
+//
+// It provides information about Kibana assets and a single source of truth for what things are legacy and should no longer be used.
+package kbncontent
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/stretchr/objx"
+)
+
+// VisualizationDescriptor describes a visualization.
+// In this case, "visualization" means anything which can be embedded in a dashboard.
+type VisualizationDescriptor struct {
+	Doc             map[string]interface{}
+	SavedObjectType string
+	Link            string
+}
+
+func (v VisualizationDescriptor) findDocumentPathsAsString(paths []string) string {
+	m := objx.Map(v.Doc)
+
+	for _, path := range paths {
+		if m.Get(path).IsStr() {
+			return m.Get(path).Str()
+		}
+	}
+
+	return ""
+}
+
+// Type returns the root-level visualization type
+// currently empty for Lens
+func (v VisualizationDescriptor) Type() string {
+	if v.SavedObjectType != "visualization" {
+		return ""
+	}
+
+	return v.findDocumentPathsAsString([]string{
+		"attributes.type",
+		"attributes.visState.type",
+		"embeddableConfig.savedVis.type", // by-value dashboard panel
+	})
+}
+
+// Editor returns the name of the visualization editor
+func (v VisualizationDescriptor) Editor() string {
+	switch v.SavedObjectType {
+	case "lens":
+		return "Lens"
+	case "map":
+		return "Maps"
+	case "search":
+		return "Discover"
+	case "visualization":
+		switch v.Type() {
+		case "metrics":
+			return "TSVB"
+		case "vega":
+			return "Vega"
+		case "timelion":
+			return "Timelion"
+		default:
+			return "Aggs-based"
+		}
+	}
+	return "Unknown"
```

Comment:
Same idea here, not sure but this seems like an unlikely edge case and would be nice to surface an error if the type is not recognized.